### ISSUE TITLE
Clarification on `OTEL_SDK_DISABLED` scope

### ISSUE THREAD

**Issue by kaylareopelle** [open-telemetry/opentelemetry-specification#4332]
**What are you trying to achieve?**

[Ruby received a PR](https://github.com/open-telemetry/opentelemetry-ruby/pull/1773) to implement the `OTEL_SDK_DISABLED` environment variable, but we're not sure how far the functionality needs to go. The PR as it stands would skip creating providers through configuration, but someone could still manually create functional providers and tracers/meters/loggers. Is that acceptable?

**Additional context.**

[Here's the description:](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#general-sdk-configuration)
> Boolean value. If “true”, a no-op SDK implementation will be used for all telemetry signals. Any other value or absence of the variable will have no effect and the SDK will remain enabled. This setting has no effect on propagators configured through the OTEL_PROPAGATORS variable.

I took a look at https://github.com/open-telemetry/opentelemetry-python/pull/3648, and it seems like they always return no-ops when `OTEL_SDK_DISABLED` is true.

However, [Java's implementation](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java#L449) seems closer to the Ruby PR. Configuration is skipped, but I don't know enough about Java to know if that means you cannot create working tracers, etc.

**Comment by MrAlias** [comment]
cc @open-telemetry/python-approvers @open-telemetry/java-approvers

**Comment by trask** [comment]
> However, [Java's implementation](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java#L449) seems closer to the Ruby PR. Configuration is skipped, but I don't know enough about Java to know if that means you cannot create working tracers, etc.

If you are using the Java autoconfigure module and set `OTEL_SDK_DISABLED=true` then none of the autoconfigure callbacks to customize the OpenTelemetry SDK are called, and so there's no way to add your own tracers etc.

**Comment by kaylareopelle** [comment]
> If you are using the Java autoconfigure module and set `OTEL_SDK_DISABLED=true` then none of the autoconfigure callbacks to customize the OpenTelemetry SDK are called, and so there's no way to add your own tracers etc. 

Thanks for clarifying, @trask. Is there a way for to use the Java SDK without the autoconfigure module? If so, does `OTEL_SDK_DISABLED` have any impact on that path?

**Comment by trask** [comment]
ya, you can use the Java SDK without the autoconfigure module, in which case `OTEL_SDK_DISABLED` has no effect

**Comment by danielgblanco** [comment]
@trask thinking about your response, and reading the spec, should `OTEL_SDK_DISABLED` also have effect with custom SDK config (i.e. without autoconfigure)?

The way I read it

>a no-op SDK implementation will be used for all telemetry signals

I understand that if I pass `OTEL_SDK_DISABLED` then a no-op SDK would override any other config?

**Comment by trask** [comment]
> @trask thinking about your response, and reading the spec, should `OTEL_SDK_DISABLED` also have effect with custom SDK config (i.e. without autoconfigure)?

I think this falls under (**emphasis added**) https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#implementation-guidelines:

> Environment variables MAY be handled (implemented) directly by a component, in the SDK, **or in a separate component (e.g. environment-based autoconfiguration component)**.

**Comment by xrmx** [comment]
> Thanks, @trask!
> 
> @open-telemetry/python-approvers - Am I reading your code correctly, will `OTEL_SDK_DISABLED` create no-ops for all attempts at creating tracer providers, etc.?

It return no-ops when calling `get_logger`, `get_tracer` and `get_meter` of the sdk signals providers, so it's not a no-op tracer provider but a no-op tracer.

**Comment by kaylareopelle** [comment]
> I think this falls under (emphasis added) https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#implementation-guidelines:
>
> > Environment variables MAY be handled (implemented) directly by a component, in the SDK, **or in a separate component (e.g. environment-based autoconfiguration component)**.

@trask - I'm not sure I understand how this answers [@danielgblanco's question](https://github.com/open-telemetry/opentelemetry-specification/issues/4332#issuecomment-2572843450). Does this mean an implementation can choose to apply an environment variable configuration only in an autoconfiguration component and have no impact on lower-level ways to access the SDK?

**Comment by trask** [comment]
> Does this mean an implementation can choose to apply an environment variable configuration only in an autoconfiguration component and have no impact on lower-level ways to access the SDK?

yeah, this is how Java SDK works. the SDK itself has no knowledge of env vars. you have to use a separate autoconfigure module in order to create an instance of the SDK based on env vars

there are definitely pros and cons of this approach, I'd probably look at a couple of other language implementations and consider what would be best specifically for Ruby

also keep in mind that future otel configuration will be built on top of [declarative config](https://github.com/open-telemetry/opentelemetry-configuration) in case that has any affect on your thinking

**Comment by aka14048** [comment]
Hi, when I tried `OTEL_SDK_DISABLED=true`
My spring application <ins>was still sending traces to Jaeger </ins>

> [!NOTE]
> (My test setup is Springboot application sending traces to Jaeger directly as provided in OTEL sample project
> LFS148-code/exercises/automatic-instrumentation/initial/todobackend-springboot/src/main/java/io/novatec/todobackend/TodobackendApplication.java)

Below is the sample code that I used
```java
Resource resource = Resource.getDefault().toBuilder().put(ServiceAttributes.SERVICE_NAME, "my-spring-service").build();

OtlpGrpcSpanExporter spanExporter = OtlpGrpcSpanExporter.builder()
.setEndpoint("http://localhost:4317") // or change to Jaeger OTLP endpoint
.build();

SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
          .addSpanProcessor(BatchSpanProcessor.builder(spanExporter).build())
          .setResource(Resource.getDefault().merge(resource))
          .build();

OpenTelemetry openTelemetry = OpenTelemetrySdk.builder().setTracerProvider(sdkTracerProvider)
      .setMeterProvider(sdkMeterProvider)
      .buildAndRegisterGlobal();

TracerProvider tracerProvider = openTelemetry.getTracerProvider();
Tracer tracer = tracerProvider.tracerBuilder("SpringBoot-Backend-App").setInstrumentationVersion("1.0")
            .setSchemaUrl("https://example")
            .build();
Span span = tracer.spanBuilder("addTodo").setSpanKind(SpanKind.INTERNAL).startSpan();
```

**Comment by trask** [comment]
hi @aka14048, please ask your question in https://github.com/open-telemetry/opentelemetry-java, thanks!

**Comment by svrnm** [comment]
closing this as stalled, the `triage:followup` got into the way, but we close stalled issues after 14 days, if this is incorrect re-open or raise new, thanks
